### PR TITLE
feat(object store): introduce object prefix for opendal object store

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
This PR introduce a new system params `use_new_object_prefix_strategy` to control whether split object prefix. In general,
- if it's a new cluster, the `use_new_object_prefix_strategy` is set to true and all opendal object store will split prefix;
- if it's an old cluster, the `use_new_object_prefix_strategy` is set to false and everything keep the same with before.

Thus, we can guarantee backward compatibility.

Once an origin customer find some issue with the performance and want to split prefix for his old cluster, it's required to execute a script to completely rename the data in the object store and modify the system params in some way.

So in a word, this pr **will not** affect the behavior of an already exist cluster.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by wcy-fdu** [comment]
PTAL❤️cc @hzxa21 @Li0k

File: e2e_test/batch/catalog/pg_settings.slt.part [hunk 1]
```diff
@@ -14,6 +14,7 @@ postmaster checkpoint_frequency
 postmaster enable_tracing
 postmaster max_concurrent_creating_streaming_jobs
 postmaster pause_on_next_bootstrap
+postmaster use_new_object_prefix_strategy
```
Comment (hzxa21) [inline e2e_test/batch/catalog/pg_settings.slt.part:17]: Is `use_new_object_prefix_strategy` mutable? I think it is immutable because changing it requires manual intervention to migrate data. We should make it `internal` according to https://github.com/risingwavelabs/risingwave/blob/380912047757c25628e82c69e47a7fe8ef7a4c6d/src/frontend/src/catalog/system_catalog/pg_catalog/pg_settings.rs#L38.

File: src/common/src/system_param/mod.rs [hunk 2]
```diff
@@ -87,6 +87,7 @@ macro_rules! for_all_params {
             { max_concurrent_creating_streaming_jobs,   u32,    Some(1_u32),                                true,   "Max number of concurrent creating streaming jobs.", },
             { pause_on_next_bootstrap,                  bool,   Some(false),                                true,   "Whether to pause all data sources on next bootstrap.", },
             { enable_tracing,                           bool,   Some(false),                                true,   "Whether to enable distributed tracing.", },
+            { use_new_object_prefix_strategy,           bool,   Some(false),                                true,   "Whether to split object prefix.", },
```
Comment (hzxa21) [inline src/common/src/system_param/mod.rs:90]: ditto: mutable should be false.

File: src/common/src/system_param/mod.rs [hunk 3]
```diff
@@ -441,6 +442,7 @@ mod tests {
             (MAX_CONCURRENT_CREATING_STREAMING_JOBS_KEY, "1"),
             (PAUSE_ON_NEXT_BOOTSTRAP_KEY, "false"),
             (ENABLE_TRACING_KEY, "true"),
+            (USE_NEW_OBJECT_PREFIX_STRATEGY_KEY, "true"),
```
Comment (hzxa21) [inline src/common/src/system_param/mod.rs:445]: ditto

File: src/jni_core/src/hummock_iterator.rs [hunk 4]
```diff
@@ -105,6 +105,7 @@ impl HummockJavaBindingIterator {
             state_store_metrics: Arc::new(global_hummock_state_store_metrics(
                 MetricLevel::Disabled,
             )),
+            use_new_object_prefix_strategy: false,
```
Comment (hzxa21) [inline src/jni_core/src/hummock_iterator.rs:108]: Let's make `use_new_object_prefix_strategy` part of `ReadPlan` instead of hardcoding it to false.

File: src/meta/src/hummock/manager/mod.rs [hunk 5]
```diff
@@ -224,13 +224,39 @@ impl HummockManager {
...
+                            .await
+                            .unwrap();
+                    }
+                    SystemParamsManagerImpl::Sql(mgr) => {
+                        mgr.set_param("use_new_object_prefix_strategy", Some("false".to_owned()))
+                            .await
+                            .unwrap();
+                    }
+                };
+            }
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/mod.rs:259]: I suggest we move the logic to [`MetaSrvEnv::new`](https://github.com/risingwavelabs/risingwave/blob/380912047757c25628e82c69e47a7fe8ef7a4c6d/src/meta/src/manager/env.rs#L361) instead here becuase:
- `cluster_first_launch` can be obtained directly there without interacting with object store and modifying `write_exclusive_cluster_id`
- We can avoid making `use_new_object_prefix_strategy` mutable and instead pass the init value for `use_new_object_prefix_strategy` to init system params manager.

File: src/object_store/src/object/mod.rs [hunk 6]
```diff
@@ -325,8 +325,25 @@ impl ObjectStoreImpl {
...
+                .inner
+                .get_object_prefix(obj_id, use_new_object_prefix_strategy),
+            ObjectStoreImpl::S3(store) => store
+                .inner
+                .get_object_prefix(obj_id, use_new_object_prefix_strategy),
+            #[cfg(madsim)]
+            ObjectStoreImpl::Sim(store) => store
+                .inner
+                .get_object_prefix(obj_id, use_new_object_prefix_strategy),
+        }
```
Comment (hzxa21) [inline src/object_store/src/object/mod.rs:346]: Is this change necessary? Why can't we use the previous `dispatch_object_store_enum!` here?

File: src/object_store/src/object/opendal_engine/opendal_object_store.rs [hunk 7]
```diff
@@ -72,18 +72,20 @@ impl OpendalObjectStore {
...
-            EngineType::Azblob => String::default(),
-            EngineType::Fs => String::default(),
+            EngineType::Hdfs
+            | EngineType::Gcs
+            | EngineType::Obs
+            | EngineType::Oss
+            | EngineType::Webhdfs
+            | EngineType::Azblob
+            | EngineType::Fs => {
+                prefix::opendal_engine::get_object_prefix(obj_id, use_new_object_prefix_strategy)
```
Comment (hzxa21) [inline src/object_store/src/object/opendal_engine/opendal_object_store.rs:87]: Have we verified that `Fs` and `Hdfs` worked after this change? I am not sure they will auto create the prefix dir for us.

File: src/object_store/src/object/s3.rs [hunk 8]
```diff
@@ -402,8 +402,9 @@ pub struct S3ObjectStore {
 impl ObjectStore for S3ObjectStore {
     type StreamingUploader = S3StreamingUploader;
 
-    fn get_object_prefix(&self, obj_id: u64) -> String {
+    fn get_object_prefix(&self, obj_id: u64, _use_new_object_prefix_strategy: bool) -> String {
         // Delegate to static method to avoid creating an `S3ObjectStore` in unit test.
+        // Using aws s3 sdk as object storage, the object prefix will be devised by default.
```
Comment (hzxa21) [inline src/object_store/src/object/s3.rs:407]: typo: `devised` -> `divided`

File: src/storage/benches/bench_compactor.rs [hunk 9]
```diff
@@ -80,6 +80,7 @@ pub async fn mock_sstable_store() -> SstableStoreRef {
         max_prefetch_block_number: 16,
         recent_filter: None,
         state_store_metrics: Arc::new(global_hummock_state_store_metrics(MetricLevel::Disabled)),
+        use_new_object_prefix_strategy: false,
```
Comment (hzxa21) [inline src/storage/benches/bench_compactor.rs:83]: better make it true because this is a bench.

File: src/storage/benches/bench_multi_builder.rs [hunk 10]
```diff
@@ -153,6 +153,7 @@ async fn generate_sstable_store(object_store: Arc<ObjectStoreImpl>) -> Arc<Sstab
         max_prefetch_block_number: 16,
         recent_filter: None,
         state_store_metrics: Arc::new(global_hummock_state_store_metrics(MetricLevel::Disabled)),
+        use_new_object_prefix_strategy: false,
```
Comment (hzxa21) [inline src/storage/benches/bench_multi_builder.rs:156]: ditto

File: src/storage/hummock_test/src/bin/replay/main.rs [hunk 11]
```diff
@@ -132,6 +132,7 @@ async fn create_replay_hummock(r: Record, args: &Args) -> Result<impl GlobalRepl
         max_prefetch_block_number: storage_opts.max_prefetch_block_number,
         recent_filter: None,
         state_store_metrics: state_store_metrics.clone(),
+        use_new_object_prefix_strategy: false,
```
Comment (hzxa21) [inline src/storage/hummock_test/src/bin/replay/main.rs:135]: let's make this part of the Args: https://github.com/risingwavelabs/risingwave/blob/380912047757c25628e82c69e47a7fe8ef7a4c6d/src/storage/hummock_test/src/bin/replay/main.rs#L55

File: src/storage/src/hummock/iterator/test_utils.rs [hunk 12]
```diff
@@ -90,6 +90,7 @@ pub async fn mock_sstable_store_with_object_store(store: ObjectStoreRef) -> Ssta
 
         recent_filter: None,
         state_store_metrics: Arc::new(global_hummock_state_store_metrics(MetricLevel::Disabled)),
+        use_new_object_prefix_strategy: false,
```
Comment (hzxa21) [inline src/storage/src/hummock/iterator/test_utils.rs:93]: Better make it true so that UT will be tested with the new strategy

File: src/tests/compaction_test/src/compaction_test_runner.rs [hunk 13]
```diff
@@ -711,6 +711,7 @@ pub async fn create_hummock_store_with_metrics(
         metrics.storage_metrics.clone(),
         metrics.compactor_metrics.clone(),
         None,
+        false,
```
Comment (hzxa21) [inline src/tests/compaction_test/src/compaction_test_runner.rs:714]: better make it true

File: e2e_test/batch/catalog/pg_settings.slt.part (hunk 1 above)
Comment (wcy-fdu) [reply e2e_test/batch/catalog/pg_settings.slt.part:17]: If it's immutable, how can we change it if we want to make an old cluster split prefix?

Comment (hzxa21) [reply e2e_test/batch/catalog/pg_settings.slt.part:17]: The mutable here means whether user can change the parameter and whether the change can take effect, not whether we can change it through manual intervention or directly modifying meta store. For example, `data_directory` is immutable, meaning that user cannot change this parameter and even if user can, it will not work. This is the same for `use_new_object_prefix_strategy`

Comment (hzxa21) [reply e2e_test/batch/catalog/pg_settings.slt.part:17]: Making the `use_new_object_prefix_strategy` mutable will be dangerous because user can techinically change it via ALTER in SQL and the after a restart, the whole cluster will crash.

Comment (wcy-fdu) [reply e2e_test/batch/catalog/pg_settings.slt.part:17]: I see.

File: src/meta/src/hummock/manager/mod.rs (hunk 5 above)
Comment (wcy-fdu) [reply src/meta/src/hummock/manager/mod.rs:259]: > cluster_first_launch can be obtained directly there

For `KV` meta store, it is, but for `SQL` meta store, can we know whether it is the first launch here?

Comment (wcy-fdu) [reply src/meta/src/hummock/manager/mod.rs:259]: cc @yezizp2012  Can we also judge cluster_first_launch for SQL backend, just like KV backend? 
Otherwise we need to modify this system variable later, but "_Making the use_new_object_prefix_strategy mutable will be dangerous because user can techinically change it via ALTER in SQL and the after a restart, the whole cluster will crash._"

File: src/object_store/src/object/opendal_engine/opendal_object_store.rs (hunk 7 above)
Comment (wcy-fdu) [reply src/object_store/src/object/opendal_engine/opendal_object_store.rs:87]: I have verified that `fs` backend work.

File: src/meta/src/hummock/manager/mod.rs (hunk 5 above)
Comment (wcy-fdu) [reply src/meta/src/hummock/manager/mod.rs]: After some offline discussion, for SQL backend I use `MigratorTrait::get_applied_migrations()` to check whether it's a new cluster:
- if `m20230908_072257_init ` is applied, it's an old cluster
- if `m20230908_072257_init ` is not applied, it's a new cluster.

Then we can make `use_new_object_prefix_strategy` immutable.

File: src/meta/src/manager/env.rs [hunk 14]
```diff
@@ -416,11 +417,25 @@ impl MetaSrvEnv {
                     .await?
                     .map(|c| c.cluster_id.to_string().into())
                     .unwrap();
+                let migrations = Migrator::get_applied_migrations(&sql_meta_store.conn).await?;
+                let mut cluster_first_launch = true;
```
Comment (wcy-fdu) [inline src/meta/src/manager/env.rs:421]: @yezizp2012 PTAL for this part, thanks!

Comment (yezizp2012) [reply src/meta/src/manager/env.rs:421]: We should check the status before calling `Migrator::up`. At this point all migrations has already been applied. Please add a helper function to get whether the cluster is first launched, together with some necessary comments.

File: src/ctl/src/common/hummock_service.rs [hunk 15]
```diff
@@ -80,11 +82,17 @@ impl HummockServiceOpts {
                 bail!(MESSAGE);
             }
         };
+        let use_new_object_prefix_strategy = match env::var("RW_USE_NEW_OBJECT_PREFIX_STRATEGY") {
+            Ok(use_new_object_prefix_strategy) => use_new_object_prefix_strategy == "true",
```
Comment (hzxa21) [inline src/ctl/src/common/hummock_service.rs:86]: I missed this in the previous review. `use_new_object_prefix_strategy` should be treated like `data_dir` because when it is incorrectly set, hummock won't work. In this case, I think we should pass `use_new_object_prefix_strategy` explicitly for `HummockServiceOpts::from_env` just like `data_dir`. This means we need to add `use_new_object_prefix_strategy` (default to true) for all ctl commands that requires `HummockServiceOpts::from_env` (in this PR you only add it for sst dump)

File: src/meta/src/manager/env.rs (hunk 14 above)
Comment (hzxa21) [reply src/meta/src/manager/env.rs:421]: > We should check the status before calling `Migrator::up`. At this point all migrations has already been applied. Please add a helper function to get whether the cluster is first launched, together with some necessary comments.

+1. If we check `get_applied_migrations` here, it will contain "m20230908_072257_init" for sure.

File: src/ctl/src/common/hummock_service.rs (hunk 15 above)
Comment (wcy-fdu) [reply src/ctl/src/common/hummock_service.rs]: That make sense, fixed by passing `use_new_object_prefix_strategy` explicitly for all ctl command.

File: src/meta/src/manager/env.rs (hunk 14 above)
Comment (wcy-fdu) [reply src/meta/src/manager/env.rs]: Fixed, and since we explicitly specify that `use_new_prefix_strategy` should be passed in, there is no need to make any further judgment for ctl,  right?

File: src/config/example.toml [hunk 16]
```diff
@@ -239,3 +239,4 @@ bloom_false_positive = 0.001
 max_concurrent_creating_streaming_jobs = 1
 pause_on_next_bootstrap = false
 enable_tracing = false
+use_new_object_prefix_strategy = false
```
Comment (hzxa21) [inline src/config/example.toml:242]: Given that `use_new_object_prefix_strategy` must be initialized, I don't think we need to include it here like what we did for `state_store` / `data_directory`.

File: src/meta/src/manager/env.rs [hunk 17]
```diff
@@ -350,6 +350,7 @@ impl MetaSrvEnv {
         init_system_params: SystemParams,
         init_session_config: SessionConfig,
         meta_store_impl: MetaStoreImpl,
+        is_sql_backend_cluster_first_launch: bool,
```
Comment (hzxa21) [inline src/meta/src/manager/env.rs:353]: Passing this only for sql backend seems a littble bit odd. Can we either move the `Migrator::up` for sql backend inside `MetaSrvEnv::new` or move the `cluster_first_launch` check for etcd metastore outside `MetaSrvEnv::new` to make the logic more consistent for both backends?

File: src/meta/src/manager/env.rs [hunk 18]
```diff
@@ -416,11 +417,17 @@ impl MetaSrvEnv {
                     .await?
                     .map(|c| c.cluster_id.to_string().into())
                     .unwrap();
+
+                let mut system_params = init_system_params;
+                // For new clusters, the name of the object store needs to be prefixed according to the object id.
+                // For old clusters, the prefix is ​​not divided for the sake of compatibility.
+                system_params.use_new_object_prefix_strategy =
+                    Some(is_sql_backend_cluster_first_launch);
```
Comment (hzxa21) [inline src/meta/src/manager/env.rs:425]: Did we forget to set `use_new_object_prefix_strategy` when etcd backend is used in L377?

File: src/common/src/system_param/mod.rs [hunk 19]
```diff
@@ -87,6 +87,7 @@ macro_rules! for_all_params {
             { max_concurrent_creating_streaming_jobs,   u32,    Some(1_u32),                                true,   "Max number of concurrent creating streaming jobs.", },
             { pause_on_next_bootstrap,                  bool,   Some(false),                                true,   "Whether to pause all data sources on next bootstrap.", },
             { enable_tracing,                           bool,   Some(false),                                true,   "Whether to enable distributed tracing.", },
+            { use_new_object_prefix_strategy,           bool,   Some(false),                                false,   "Whether to split object prefix.", },
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/common/src/system_param/mod.rs:90]: Make sense.

File: src/meta/src/manager/env.rs (hunk 18 above)
Comment (wcy-fdu) [reply src/meta/src/manager/env.rs:425]: No, for etcd backend, aka SQL backend, we will use the origin variable cluster_first_launch as the `use_new_object_prefix_strategy`, see L71 in `system_param/mod.rs`.
```
 params.use_new_object_prefix_strategy = Some(cluster_first_launch);
```

**Comment by wcy-fdu** [comment]
Backward compatibility tests pass.

**
image
**

File: src/meta/src/manager/env.rs [hunk 20]
```diff
@@ -344,6 +345,26 @@ impl MetaOpts {
...
+pub async fn is_first_launch_for_sql_backend_cluster(
+    sql_meta_store: &SqlMetaStore,
+) -> MetaResult<bool> {
+    let migrations = Migrator::get_applied_migrations(&sql_meta_store.conn).await?;
+    let mut cluster_first_launch = true;
+    for migration in migrations {
+        if migration.name() == "m20230908_072257_init"
+            && migration.status() == MigrationStatus::Applied
+        {
+            cluster_first_launch = false;
```
Comment (hzxa21) [inline src/meta/src/manager/env.rs:362]: nits: you can return directly here and don't need to continue the iteration.

File: src/meta/src/manager/env.rs [hunk 21]
```diff
@@ -411,16 +432,31 @@ impl MetaSrvEnv {
                 }
             }
             MetaStoreImpl::Sql(sql_meta_store) => {
+                let mut is_sql_backend_cluster_first_launch = true;
+                if let MetaStoreImpl::Sql(sql_store) = &meta_store_impl {
+                    is_sql_backend_cluster_first_launch =
+                        is_first_launch_for_sql_backend_cluster(sql_store).await?;
+                    // Try to upgrade if any new model changes are added.
+                    Migrator::up(&sql_store.conn, None)
```
Comment (hzxa21) [inline src/meta/src/manager/env.rs:440]: In L376, `meta_store_impl` is passed to new `NotificationManager`. Have we checked that whether it is correct to init `NotificationManager` before `Migrator::up` ?

File: src/meta/src/manager/env.rs (hunk 18 above)
Comment (hzxa21) [reply src/meta/src/manager/env.rs:454]: This is incorrect. Consider the following case:
1. Cluster is created with the new image after this PR.
2. `cluster_first_launch` is true on meta pod starts up for the first time and `use_new_object_prefix_strategy` is persisted to be true in etcd.
3. meta pod restarts
4. `cluster_first_launch` is false this time, but we will still hit L71 in `system_param/mod.rs`, resulting in `use_new_object_prefix_strategy` mistakenly changed to false.

This is all because we change the system params value after `merge_params` in `SystemParamsManager::new`. I think we should following the semantic of `init_params` exactly and provide the initial value of `use_new_object_prefix_strategy` in `init_params` here outside of `SystemParamsManager::new`

**Review by hzxa21 — APPROVED** [review]
LGTM. Thanks for the PR!

File: src/meta/node/src/server.rs [hunk 22]
```diff
@@ -408,13 +407,6 @@ pub async fn start_service_as_election_leader(
     mut svc_shutdown_rx: WatchReceiver<()>,
 ) -> MetaResult<()> {
     tracing::info!("Defining leader services");
-    if let MetaStoreImpl::Sql(sql_store) = &meta_store_impl {
```
Comment (Li0k) [inline src/meta/node/src/server.rs:411]: why remove this code ?
